### PR TITLE
feat: add filter, sort & search functionality

### PR DESCRIPTION
**Pull request by PuneetP16**
Features added: 
1. filter by category (from homescreen and on explore page)
2. sort by date
3. search from explore and home page, redirects to the explore page with result.

In desktop mode: search is embedded in header 
In mobile mode: search is kept as independent component

image

image

image

on Clicking on All button the all the filter and search gets reset.

### CODE REVIEW

File: src/components/Cards/CategoryCard/CategoryCard.jsx [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
-				// 		type: "CATEGORY_SELECTION",
-				// 		payload: prodCat.categoryName,
-				// 	})
-				// }
+				onClick={() => {
+					filterDispatch({
+						type: "CATEGORY",
+						payload: category.categoryId,
+					});
+					console.log("sadf", category.categoryId);
```
Comment (tarunsankhla) [inline src/components/Cards/CategoryCard/CategoryCard.jsx:16]: Remove console

File: src/components/Cards/VideoCard/VideoCard.jsx [hunk 2]
```diff
@@ -45,18 +45,18 @@ export const VideoCard = ({ video }) => {
...
-		channelTitle,
-		publishedAt,
-		tags,
-		thumbnailURL,
-		viewCount,
-	});
+	// console.log({
+	// 	_id,
+	// 	title,
+	// 	description,
```
Comment (tarunsankhla) [inline src/components/Cards/VideoCard/VideoCard.jsx:51]: Remove comments

File: src/components/CategoryListing/CategoryListing.jsx [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import { CategoryCard } from "../Cards/CategoryCard/CategoryCard";
 
 export const CategoryListing = ({ categories }) => {
+
 	return categories.length > 0
 		? categories.map((category) => {
```
Comment (tarunsankhla) [inline src/components/CategoryListing/CategoryListing.jsx:6]: Why you need return remove the curly brace

File: src/components/Filter/Filter.jsx [hunk 4]
```diff
@@ -1,28 +1,83 @@
...
+
 	return (
 		<form className="filter_section" onSubmit={(e) => e.preventDefault()}>
 			<div className="filter_container">
-				<button className="filter_button btn--outline--primary">All</button>
+				<button
+					onClick={() => {
+						filterDispatch({
+							type: "CATEGORY",
+							payload: "All",
```
Comment (tarunsankhla) [inline src/components/Filter/Filter.jsx:19]: Can you move the logic out of html into js function

File: src/components/Filter/Filter.jsx [hunk 5]
```diff
@@ -1,28 +1,83 @@
...
 					  })
 					: null}
-				<button className="filter_button sort_button btn--outline--primary">
-					Sort by Date
+				<button
+					className={`filter_button sort_button ${
+						byLatest === "unsort"
+							? "btn--outline--primary"
+							: byLatest
+							? "btn--primary"
```
Comment (tarunsankhla) [inline src/components/Filter/Filter.jsx:56]: Try moving this out to some variable

File: src/contexts/magnificentContext.js [hunk 6]
```diff
@@ -16,19 +17,21 @@ export const MagnificentProvider = ({ children }) => {
 	return (
```
Comment (tarunsankhla) [inline src/contexts/magnificentContext.js:17]: What is this value you need your are passing a string that too empty?

**Review by tarunsankhla — APPROVED** [review]
Overall great minor changes

File: src/components/CategoryListing/CategoryListing.jsx (hunk 3 above)
Comment (PuneetP16) [reply src/components/CategoryListing/CategoryListing.jsx:6]: ok. Updated it.

File: src/components/Filter/Filter.jsx (hunk 4 above)
Comment (PuneetP16) [reply src/components/Filter/Filter.jsx:19]: Yeah will do that.

File: src/contexts/magnificentContext.js (hunk 6 above)
Comment (PuneetP16) [reply src/contexts/magnificentContext.js:17]: Here, i am passing App component.

File: src/components/Cards/VideoCard/VideoCard.jsx [hunk 7]
```diff
@@ -45,18 +45,18 @@ export const VideoCard = ({ video }) => {
...
+	// 	categoryId,
+	// 	channelId,
+	// 	channelTitle,
+	// 	publishedAt,
+	// 	tags,
+	// 	thumbnailURL,
+	// 	viewCount,
+	// });
```
Comment (dev-ashishsethi) [inline src/components/Cards/VideoCard/VideoCard.jsx:59]: remove comments

File: src/components/Filter/Filter.css [hunk 8]
```diff
@@ -1,12 +1,12 @@
 .filter_section {
-	width: 60%;
+	width: 75%;
 	margin-bottom: 5px;
 }
 
 .filter_container {
 	display: flex;
 	flex-flow: row wrap;
```
Comment (dev-ashishsethi) [inline src/components/Filter/Filter.css:8]: short hand properties have compatibility issues with some browsers

File: src/contexts/magnificentContext.js [hunk 9]
```diff
@@ -16,19 +17,21 @@ export const MagnificentProvider = ({ children }) => {
...
+									<UserProvider>
+										<ThemeProvider>
+											<LoaderProvider>{children}</LoaderProvider>
+										</ThemeProvider>
+									</UserProvider>
+								</VideoProvider>
+							</AuthProvider>
+						</ScrollToTopProvider>
+					</LoaderProvider>
+				</FilterProvider>
```
Comment (dev-ashishsethi) [inline src/contexts/magnificentContext.js:34]: make a single provider

File: src/utilities/filterOperations.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+export const sortItByDate = (list, byLatest) => {
+	if (byLatest === "unsort") return list;
+	const data = [...list];
+
+	const ascending = (a, b) => {
+		console.log(a);
```
Comment (dev-ashishsethi) [inline src/utilities/filterOperations.js:6]: remove console.log

**Review by dev-ashishsethi — APPROVED** [review]
do check comments

File: src/components/Filter/Filter.css (hunk 8 above)
Comment (PuneetP16) [reply src/components/Filter/Filter.css:8]: noted

File: src/contexts/magnificentContext.js (hunk 9 above)
Comment (PuneetP16) [reply src/contexts/magnificentContext.js:34]: It is already a single provider.

File: src/utilities/filterOperations.js (hunk 10 above)
Comment (PuneetP16) [reply src/utilities/filterOperations.js:6]: will do that

**Pull request merged**